models/locked_resource.go

```diff
@@ -0,0 +1,110 @@
+// Copyright 2020 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package models
+
+import (
+	"fmt"
+
+	"code.gitea.io/gitea/modules/setting"
+)
+
+// LockedResource represents the locking key for a pessimistic
+// lock that can hold a counter
+type LockedResource struct {
+	LockType string `xorm:"pk VARCHAR(30)"`
+	LockKey  int64  `xorm:"pk"`
+	Counter  int64  `xorm:"NOT NULL DEFAULT 0"`
+}
+
+// GetLockedResource gets or creates a pessimistic lock on the given type and key
+func GetLockedResource(e Engine, lockType string, lockKey int64) (*LockedResource, error) {
+	locked := &LockedResource{LockType: lockType, LockKey: lockKey}
+
+	if err := upsertLockedResource(e, locked); err != nil {
+		return nil, fmt.Errorf("upsertLockedResource: %v", err)
+	}
+
+	// Read back the record we've created or locked to get the current Counter value
+	if has, err := e.Table(locked).Get(locked); err != nil {
+		return nil, fmt.Errorf("get locked resource %s:%d: %v", lockType, lockKey, err)
+	} else if !has {
+		return nil, fmt.Errorf("unexpected upsert fail  %s:%d", lockType, lockKey)
+	}
+
+	return locked, nil
+}
+
+// UpdateLockedResource updates the value of the counter of a locked resource
+func UpdateLockedResource(e Engine, resource *LockedResource) error {
+	_, err := e.Table(resource).Cols("counter").Update(resource)
+	return err
+}
+
+// DeleteLockedResource deletes a locked resource
+func DeleteLockedResource(e Engine, resource *LockedResource) error {
+	_, err := e.Delete(resource)
+	return err
+}
+
+// TempLockResource locks the given key but does not leave a permanent record
+func TempLockResource(e Engine, lockType string, lockKey int64) error {
```

Comment:
This is basically a (key)notexists or (key)isavailable test. That might be a better name depending on how you were intending on using this.

Reply:
The intention is however to _lock_ a type of resource. The "(key)notexists functionality" is only a convenience to help preventing a bad usage. Maybe I should  rename it to something_something_Mutex?

Reply:
Changed the name to a more informative but verbose `TemporarilyLockResourceKey`.